src/hipscat_import/catalog/margin_cache.py

```diff
@@ -0,0 +1,249 @@
+                margin_order=args.margin_order,
+                ra_column=args.catalog.catalog_info.ra_column,
+                dec_column=args.catalog.catalog_info.dec_column,
+            )
+        )
+
+    for _ in tqdm(
+        as_completed(futures, with_results=True),
+        desc="Mapping  ",
+        total=len(futures),
+    ):
+        ...
+
+def _map_pixel_shards(
+    partition_file,
+    margin_pairs,
+    margin_threshold,
+    output_path,
+    margin_order,
+    ra_column,
+    dec_column
+):
+    """Creates margin cache shards from a source partition file."""
+    data = pd.read_parquet(partition_file)
+
+    data["margin_pixel"] = hp.ang2pix(
+        2**margin_order,
+        data[ra_column].values,
+        data[dec_column].values,
+        lonlat=True,
+        nest=True,
+    )
+
+    constrained_data = data.merge(margin_pairs, on="margin_pixel")
+
+    constrained_data.groupby(["partition_order", "partition_pixel"]).apply(
+        _to_pixel_shard,
+        margin_threshold=margin_threshold,
+        output_path=output_path,
+        margin_order=margin_order,
+        ra_column=ra_column,
+        dec_column=dec_column,
+    )
+
+def _to_pixel_shard(data, margin_threshold, output_path, margin_order, ra_column, dec_column):
+    """Do boundary checking for the cached partition and then output remaining data."""
+    order, pix = data["partition_order"].iloc[0], data["partition_pixel"].iloc[0]
+    source_order, source_pix = data["Norder"].iloc[0], data["Npix"].iloc[0]
+
+    scale = pixel_math.get_margin_scale(order, margin_threshold)
+    bounding_polygons = pixel_math.get_margin_bounds_and_wcs(order, pix, scale)
+    is_polar, pole = pixel_math.pixel_is_polar(order, pix)
+
+    if is_polar:
+        data = _margin_filter_polar(
+            data,
+            order,
+            pix,
+            margin_order,
+            pole,
+            margin_threshold,
+            ra_column,
+            dec_column,
+            bounding_polygons,
+        )
+    else:
+        data["margin_check"] = pixel_math.check_margin_bounds(
+            data[ra_column].values,
+            data[dec_column].values,
+            bounding_polygons
+        )
+
+    # pylint: disable-next=singleton-comparison
+    margin_data = data.loc[data["margin_check"] == True]
+
+    if len(margin_data):
+        # TODO: this should be a utility function in `hipscat`
+        # that properly handles the hive formatting
+        # generate a file name for our margin shard
+        partition_file = paths.pixel_catalog_file(output_path, order, pix)
+        partition_dir = f"{partition_file[:-8]}/"
+        shard_dir = paths.pixel_directory(
+            partition_dir, source_order, source_pix
+        )
+
+        file_io.make_directory(shard_dir, exist_ok=True)
+
+        shard_path = paths.pixel_catalog_file(
+            partition_dir, source_order, source_pix
+        )
+
+        margin_data.to_parquet(shard_path)
```

Comment:
Will this parquet file get written out with all the extra columns added above? Will those get stripped out when we reduce the shards together, or could we just keep these notes in a separate data frame so it's clear that we're only writing out the original fields?

Reply:
good catch I'll remove them!